explicitly access Container.State instead of through embedded struct

thaJeztah opened the pull request:
- [x] stacked on https://github.com/moby/moby/pull/50767

### explicitly access Container.State instead of through embedded struct

The Container.State struct holds the container's state, and most of its fields are expected to change dynamically. Some o these state-changes are explicit, for example, setting the container to be "stopped". Other state changes can be more explicit, for example due to the containers' process exiting or being "OOM" killed by the kernel.

The distinction between explicit ("desired") state changes and "state" ("actual state") is sometimes vague; for some properties, we clearly separated them, for example if a user requested the container to be stopped or restarted, we store state in the Container object itself;

    HasBeenManuallyStopped   bool // used for unless-stopped restart policy
    HasBeenManuallyRestarted bool `json:"-"` // used to distinguish restart caused by restart policy from the manual one

Other properties are more ambiguous. such as "HasBeenStartedBefore" and "RestartCount", which are stored on the Container (and persisted to disk), but may be more related to "actual" state, and likely should not be persisted;

    RestartCount             int
    HasBeenStartedBefore     bool

Given that (per the above) concurrency must be taken into account, most changes to the `container.State` struct should be protected; here's where things get blurry. While the `State` type provides various accessor methods, only some of them take concurrency into account; for example, [State.IsRunning] and [State.GetPID] acquire a lock, whereas [State.ExitCodeValue] does not. Even the (commonly used) [State.StateString] has no locking at all.

The way to handle this is error-prone; [container.State] contains a mutex, and it's exported. Given that its embedded in the [container.Container] struct, it's also exposed as an exported mutex for the container. The assumption here is that by "merging" the two, the caller to acquire a lock when either the container _or_ its state must be mutated. However, because some methods on `container.State` handle their own locking, consumers must be deeply familiar with the internals; if both changes to the `Container` AND `Container.State` must be made. This gets amplified more as some (exported!) methods, such as [container.SetRunning] mutate multiple fields, but don't acquire a lock (so expect the caller to hold one), but their (also exported) counterpart (e.g. [State.IsRunning]) do.

It should be clear from the above, that this needs some architectural changes; a clearer separation between "desired" and "actual" state (opening the potential to update the container's config without manually touching its `State`), possibly a method to obtain a read-only copy of the current state (for those querying state), and reviewing which fields belong where (and should be persisted to disk, or only remain in memory).

This PR preserves the status quo; it makes no structural changes, other than exposing where we access the container's state. Where previously the State fields and methods were referred to as "part of the container" (e.g. `ctr.IsRunning()` or `ctr.Running`), we now explicitly reference the embedded `State` (`ctr.State.IsRunning`, `ctr.State.Running`).

The exception (for now) is the mutex, which is still referenced through the embedded struct (`ctr.Lock()` instead of `ctr.State.Lock()`), as this is (mostly) by design to protect the container, and what's in it (including its `State`).

[State.IsRunning]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L205-L209
[State.GetPID]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L211-L216
[State.ExitCodeValue]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L218-L228
[State.StateString]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L102-L131
[container.State]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L15-L23
[container.Container]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/container.go#L67-L75
[container.SetRunning]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L230-L277


### daemon/container: remove NewState() constructor

This constructor did not do anything other than creating an empty struct
for an exported type. While we should look at initializing with a proper
state, we currently do not, so let's not pretend we do some magic here,
and leave it for a future exercise to create a proper constructor if we
need one.

### daemon/container: remove State.ExitCode() method

This method did not provide any special handling for accessing the
field, and did not handle locking. Let's remove it for now to
not pretend we're doing anything more safe than directly accessing
the field.

### daemon/container: remove State.SetExitCode() method

This method did not provide any special handling for accessing the
field, and did not handle locking. Let's remove it for now to
not pretend we're doing anything more safe than directly accessing
the field.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog

```

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
explicitly access Container.State instead of through embedded struct

The Container.State struct holds the container's state, and most of
its fields are expected to change dynamically. Some o these state-changes
are explicit, for example, setting the container to be "stopped". Other
state changes can be more explicit, for example due to the containers'
process exiting or being "OOM" killed by the kernel.

The distinction between explicit ("desired") state changes and "state"
("actual state") is sometimes vague; for some properties, we clearly
separated them, for example if a user requested the container to be
stopped or restarted, we store state in the Container object itself;

    HasBeenManuallyStopped   bool // used for unless-stopped restart policy
    HasBeenManuallyRestarted bool `json:"-"` // used to distinguish restart caused by restart policy from the manual one

Other properties are more ambiguous. such as "HasBeenStartedBefore" and
"RestartCount", which are stored on the Container (and persisted to
disk), but may be more related to "actual" state, and likely should
not be persisted;

    RestartCount             int
    HasBeenStartedBefore     bool

Given that (per the above) concurrency must be taken into account, most
changes to the `container.State` struct should be protected; here's where
things get blurry. While the `State` type provides various accessor methods,
only some of them take concurrency into account; for example, [State.IsRunning]
and [State.GetPID] acquire a lock, whereas [State.ExitCodeValue] does not.
Even the (commonly used) [State.StateString] has no locking at all.

The way to handle this is error-prone; [container.State] contains a mutex,
and it's exported. Given that its embedded in the [container.Container]
struct, it's also exposed as an exported mutex for the container. The
assumption here is that by "merging" the two, the caller to acquire a lock
when either the container _or_ its state must be mutated. However, because
some methods on `container.State` handle their own locking, consumers must
be deeply familiar with the internals; if both changes to the `Container`
AND `Container.State` must be made. This gets amplified more as some
(exported!) methods, such as [container.SetRunning] mutate multiple fields,
but don't acquire a lock (so expect the caller to hold one), but their
(also exported) counterpart (e.g. [State.IsRunning]) do.

It should be clear from the above, that this needs some architectural
changes; a clearer separation between "desired" and "actual" state (opening
the potential to update the container's config without manually touching
its `State`), possibly a method to obtain a read-only copy of the current
state (for those querying state), and reviewing which fields belong where
(and should be persisted to disk, or only remain in memory).

This PR preserves the status quo; it makes no structural changes, other
than exposing where we access the container's state. Where previously the
State fields and methods were referred to as "part of the container"
(e.g. `ctr.IsRunning()` or `ctr.Running`), we now explicitly reference
the embedded `State` (`ctr.State.IsRunning`, `ctr.State.Running`).

The exception (for now) is the mutex, which is still referenced through
the embedded struct (`ctr.Lock()` instead of `ctr.State.Lock()`), as this
is (mostly) by design to protect the container, and what's in it (including
its `State`).

[State.IsRunning]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L205-L209
[State.GetPID]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L211-L216
[State.ExitCodeValue]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L218-L228
[State.StateString]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L102-L131
[container.State]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L15-L23
[container.Container]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/container.go#L67-L75
[container.SetRunning]: https://github.com/moby/moby/blob/c4afa7715715a1020e50b19ad60728c4479fb0a5/daemon/container/state.go#L230-L277

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon/container: remove NewState() constructor

This constructor did not do anything other than creating an empty struct
for an exported type. While we should look at initializing with a proper
state, we currently do not, so let's not pretend we do some magic here,
and leave it for a future exercise to create a proper constructor if we
need one.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on daemon/inspect.go:
```diff
@@ -135,7 +135,7 @@ func (daemon *Daemon) getInspectData(daemonCfg *config.Config, container *contai
 		Dead:       container.State.Dead,
 		Pid:        container.State.Pid,
-		ExitCode:   container.State.ExitCode(),
+		ExitCode:   container.State.ExitCodeValue,
```
`State.ExitCode()` pretended to be an accessor, but handles no locking or otherwise, so I'm deliberately removing the function to expose that no special handling takes place; https://github.com/moby/moby/blob/f74e5d48b3a8a83a6b969be2dd2b14af3d9d4d22/daemon/container/state.go#L70-L73

thaJeztah commented on daemon/container/container.go:
```diff
@@ -534,7 +534,11 @@ func (container *Container) GetExecIDs() []string {
 func (container *Container) ShouldRestart() bool {
-	shouldRestart, _, _ := container.RestartManager().ShouldRestart(uint32(container.ExitCode()), container.HasBeenManuallyStopped, container.FinishedAt.Sub(container.StartedAt))
+	shouldRestart, _, _ := container.RestartManager().ShouldRestart(
+		uint32(container.State.ExitCodeValue),
```
Same for this one; `State.ExitCode()` pretended to be an accessor, but handles no locking or otherwise, so I'm deliberately removing the function to expose that no special handling takes place; https://github.com/moby/moby/blob/f74e5d48b3a8a83a6b969be2dd2b14af3d9d4d22/daemon/container/state.go#L70-L73

thaJeztah committed:
daemon/container: remove State.ExitCode() method

This method did not provide any special handling for accessing the
field, and did not handle locking. Let's remove it for now to
not pretend we're doing anything more safe than directly accessing
the field.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

robmry commented on daemon/container/container.go:
```diff
@@ -837,11 +841,11 @@ func (container *Container) RestoreTask(ctx context.Context, client libcontainer
 		return err
 	}
-	container.task, err = container.ctr.AttachTask(ctx, container.InitializeStdio)
+	container.State.task, err = container.ctr.AttachTask(ctx, container.InitializeStdio)
```
```suggestion
container.State.task, err = container.ctr.State.AttachTask(ctx, container.InitializeStdio)
```

robmry commented on daemon/health.go:
```diff
@@ -352,7 +352,7 @@ func (daemon *Daemon) updateHealthMonitor(c *container.Container) {
 
 	probe := getProbe(c)
-	wantRunning := c.Running && !c.Paused && probe != nil
+	wantRunning := c.State.Running && !c.State.Paused && probe != nil
```
There's another one on line 267 of this file - `monitor().getInterval()` ... `c.State.Health.Health.Status`.

thaJeztah replied:
Thx for the DM; I think I tacked the last one now as well! I had trouble finding it ❤️

thaJeztah committed:
explicitly access Container.State.Health.Health

The State.Health struct has a mutex, but in various places
we access the embedded Health struct directly.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on daemon/container/view.go:
```diff
@@ -294,14 +294,9 @@ func (v *View) GetAllNames() map[string][]string {
-		FailingStreak: failingStreak,
+	if ctr.State.Health != nil {
+		health = ctr.State.Health.Status()
+		failingStreak = ctr.State.Health.Health.FailingStreak
```
For others; I pushed an extra commit that also exposes places where we access `State.Health.Health.<field>` - `State.Health` contains a mutex, but we have various places where we access fields inside it without obtaining a lock.

_Possibly_ some of those are "ok", but trying to expose the ugliness so that we can look at those uses, and where needed make design changes to make sure it's robust.

robmry commented:
As discussed - pushed an update to the `Health.Health` change to include some more direct accesses to the embedded struct, and ugly it up some more!
